pkg/aggregator/aggregator.go

```diff
@@ -28,21 +27,42 @@ import (
 )
 
 // usedDefinitionForSpec returns a map with all used definitions in the provided spec as keys and true as values.
-func usedDefinitionForSpec(sp *spec.Swagger) map[string]bool {
+func usedDefinitionForSpec(root *spec.Swagger) map[string]bool {
 	usedDefinitions := map[string]bool{}
-	walkOnAllReferences(func(ref spec.Ref) spec.Ref {
-		if refStr := ref.String(); refStr != "" && strings.HasPrefix(refStr, definitionPrefix) {
-			usedDefinitions[refStr[len(definitionPrefix):]] = true
+	alreadyVisited := map[string]bool{}
+
+	walker := &readonlyReferenceWalker{
+		root: root,
+	}
+	walker.walkRefCallback = func(ref *spec.Ref) {
+		refStr := ref.String()
+		if refStr == "" || !strings.HasPrefix(refStr, definitionPrefix) {
+			return
 		}
-		return ref
-	}, sp)
+
+		defName := refStr[len(definitionPrefix):]
+		usedDefinitions[defName] = true
+
+		if _, found := root.Definitions[defName]; found && !alreadyVisited[refStr] {
+			alreadyVisited[refStr] = true
+			def := root.Definitions[defName]
+			walker.walkSchema(&def)
+		}
+	}
+	walker.Start()
+
 	return usedDefinitions
 }
 
 // FilterSpecByPaths removes unnecessary paths and definitions used by those paths.
 // i.e. if a Path removed by this function, all definitions used by it and not used
 // anywhere else will also be removed.
-func FilterSpecByPaths(sp *spec.Swagger, keepPathPrefixes []string) {
+// It does not modify the input, but the output shares data structures with the input.
```

Comment:
Is this a changed behaviour? `but the output shares data structures with the input.`? if it is, it would be a backward incompatible change as this function is public.

Reply:
Oh, the function has a new signature. I think it should be fine.

Reply:
sorry, goine back and forth on this. This is definitely backward incompatible. Any caller outside of the repo expect FilterSpecByPath to mutate the input.

Reply:
fixed